### PR TITLE
Area estimation colab

### PR DESCRIPTION
**Pull request by adebowaledaniel**
Colab notebook for area estimation follows the same logic as the crop area estimation notebook: https://github.com/nasaharvest/crop-mask/pull/207. The notebook is implemented according to the potential solution described in https://github.com/nasaharvest/crop-mask/issues/199; which includes:
- A [Colab notebook](https://colab.research.google.com/github/nasaharvest/crop-mask/blob/area_estimation_colab/notebooks/crop_area_estimation.ipynb) that downloads cropland mask from the cloud storage (gs://crop-mask-preds-merged) and executes the area estimation procedures,
- all crop area estimation functions moved to [src](https://github.com/nasaharvest/crop-mask/blob/e84b66276e351e2b519f12759707a6d655cbea97/src/cropareaest_funcs.py)

### CODE REVIEW

File: src/cropareaest_funcs.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
+# Import libraries
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:1]: No need for this comment

**Comment by ivanzvonkov** [comment]
@adebowaledaniel could you clear the outputs in the notebook (only the code should be version controlled) and resolve possible conflicts?

File: src/cropareaest_funcs.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+    prj_raster = [f for f in os.listdir(root_dir) if f.startswith("prj")][0]
+
+    with rio.open(prj_raster) as src:
+        bbx = src.bounds
+    geom = box(*bbx)
+    bbx_gdf = gpd.GeoDataFrame(geometry=[geom], crs=src.crs)
+    return bbx_gdf
+
+
+def load_ne(country_code, regions_of_interest):
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:25]: For all of these functions could you add type hints? 
Example:
```python
def load_ne(country_code: str, regions_of_interest: List[str]) -> gpd.GeoDataFrame:
```

This way mypy will verify that they are called with correct types.

File: src/cropareaest_funcs.py [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+from rasterio.mask import mask
+import cartopy.io.shapereader as shpreader
+from shapely.geometry import box
+from sklearn.metrics import confusion_matrix
+
+
+def return_map_bounds_as_df():
+    "Read the bounding coordinates of the crop mask"
+    root_dir = os.getcwd()
+    prj_raster = [f for f in os.listdir(root_dir) if f.startswith("prj")][0]
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:16]: Where is this `prj...` file coming from?

File: src/cropareaest_funcs.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+
+    with rio.open(prj_raster) as src:
+        bbx = src.bounds
+    geom = box(*bbx)
+    bbx_gdf = gpd.GeoDataFrame(geometry=[geom], crs=src.crs)
+    return bbx_gdf
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:22]: nit: I think this could just be one line
```python
with rio.open(prj_raster) as src:
    return gpd.GeoDataFrame(geometry=[box(*src.bounds)], crs=src.crs)
```

File: src/cropareaest_funcs.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+                regions only seen in below plot."
+            )
+            ne_gdf[condition].plot(
+                column="name", legend=True, legend_kwds={"loc": "lower right"}, figsize=(10, 10)
+            )
+        else:
+            condition = ne_gdf["name"].isin(regions_of_interest)
+            boundary = ne_gdf[condition].copy()
+            print("All regions found!")
+    return boundary
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:66]: Ideally, once functions are moved out of a notebook they should get a unittest associated with them

File: src/cropareaest_funcs.py [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+    regions_of_interest: list of regions of interest
+    """
+
+    ne_shapefile = shpreader.natural_earth(
+        resolution="10m", category="cultural", name="admin_1_states_provinces"
+    )
+    ne_gdf = gpd.read_file(ne_shapefile)
+
+    if len(regions_of_interest) == 0:
+        # region_of_interest & country_code will be collected using dropdown
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:38]: Comment is not needed here

File: src/cropareaest_funcs.py [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+            boundary = ne_gdf[condition].copy()
+            print("All regions found!")
+    return boundary
+
+
+def getFeatures(gdf):
+    """Function to parse features from GeoDataFrame in such a manner that rasterio wants them"""
+    import json
+
+    return [json.loads(gdf.to_json())["features"][0]["geometry"]]
```
Comment (ivanzvonkov) [inline src/cropareaest_funcs.py:73]: I don't think this function is necessary, the one-liner can just be used directly

**Review by ivanzvonkov — CHANGES_REQUESTED** [review]
1. I think `src/area_utils.py` would be a more fitting file name
2. Generally when functions become part of source code it would be nice for them to have associated unit tests. This can be done by creating a file `test/test_area_utils.py` and writing unit tests for each function
3. Additionally when functions become part of source code it would be nice to see them type hinted to make use of mypy checks.
4. Lastly there are some comments in the functions that don't really need to be there and can be cleaned up

File: src/cropareaest_funcs.py (hunk 3 above)
Comment (adebowaledaniel) [reply src/cropareaest_funcs.py:16]: Came from GDAL project output https://github.com/nasaharvest/crop-mask/blob/e84b66276e351e2b519f12759707a6d655cbea97/src/cropareaest_funcs.py#L131-L132
But it's removed as the function is not needed anymore.

**Comment by ivanzvonkov** [comment]
Looks like the test ran but failed on the 121-122 issue

File: src/area_utils.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+
+def random_inds(
+    binary_map: np.ndarray, strata: int, sample_size: int
+) -> Tuple[np.ndarray, np.ndarray]:
+    # """ Generate random indices for sampling from a binary map."""
+    inds = np.where(binary_map == strata)
+    rand_inds = np.random.permutation(np.arange(inds[0].shape[0]))[:sample_size]
+    rand_px = inds[0][rand_inds]
+    rand_py = inds[1][rand_inds]
+    # del inds
```
Comment (ivanzvonkov) [inline src/area_utils.py:201]: I think this is no longer needed

**Review by ivanzvonkov — APPROVED** [review]
LGTM!
